### PR TITLE
Listen to S3 file uploaded notifications and call uploadComplete on bridge

### PR DESCRIPTION
**Pull request by liujoshua**

### CODE REVIEW

File: src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+     * minutes. Package-scoped to enable unit tests to mock this out.
+     */
+    @Cacheable(lifetime = 5, unit = TimeUnit.MINUTES)
+    WorkerClient getWorkerClient() {
+        Session session = ClientProvider.signIn(credentials);
+        return session.getWorkerClient();
+    }
+
+    @Override
+    public void callback(String messageBody) throws Exception {
```
Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java:56]: If this throws an exception, shouldn't we swallow it but log it instead? Or do we know that when this throws, the encompassing framework handles logging, it doesn't kill the thread, etc.?

File: src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+        callback(notification);
+    }
+
+    // package-scoped to enable mocking/testing
+    void callback(S3EventNotification notification) {
+        notification.getRecords().stream().filter(record -> shouldProcessRecord(record)).forEach(record -> {
+            String uploadId = record.getS3().getObject().getKey();
+            LOG.info("Completing upload, id=" + uploadId);
+            getWorkerClient().completeUpload(uploadId);
```
Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java:67]: What if we flush Redis and you lose your session in the 5 minutes this is cached, what happens?

File: src/test/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallbackTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        when(record.getS3()).thenReturn(entity);
+
+        return record;
+    }
+
+    @Test
+    public void testShouldProcessRecords_S3Put() {
+        S3EventNotification.S3EventNotificationRecord s3Put = createMockRecord("aws:s3", "ObjectCreated:Put");
+
+        assert callback.shouldProcessRecord(s3Put);
```
Comment (alxdarksage) [inline src/test/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallbackTest.java]: I would prefer assertTrue()

File: src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java (hunk 1 above)
Comment (liujoshua) [inline src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java:56]: By throwing an exception, we are allowing SQS to redeliver this message.

When callback() throws an Exception, PollSqsWorker's run() method swallows it and logs the message (and skips the part where it removes the message from the queue)

Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java:56]: Okay. Does this create a loop if the error is on the Bridge side?

File: src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java (hunk 2 above)
Comment (liujoshua) [inline src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java:67]: Looks like the worker client will remain unauthorized. The message should get redelivered, which isn't ideal but we shouldn't lose a message. I'll think about better ways to authenticate

File: src/test/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallbackTest.java (hunk 3 above)
Comment (liujoshua) [inline src/test/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallbackTest.java]: Fixed

File: src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java (hunk 2 above)
Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java:67]: That's probably fine, though if the thing stays cached for 5 minutes, that could be a lot of failed messages.

File: src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java (hunk 1 above)
Comment (liujoshua) [inline src/main/java/org/sagebionetworks/bridge/exporter/notification/S3EventNotificationCallback.java:56]: In a sense. There are settings in SQS for how long messages stay invisible before qualifying for redelivery, how many times a message can be redelivered, and a maximum retention period before a message gets deleted

**Pull request merged**